Libraries/MLXLLM/LoraTrain.swift

```diff
@@ -2,222 +2,12 @@
-        scale: Float = 20.0, linear: QuantizedLinear
-    ) {
-
-        // Scale for low-rank update
-        self.scale = scale
-
-        // Low rank lora weights
-        let loraScale = 1 / sqrt(Float(inputDimensions))
-        self._loraA.wrappedValue = MLXRandom.uniform(
-            low: -loraScale, high: loraScale, [inputDimensions, rank])
-        self._loraB.wrappedValue = MLXArray.zeros([rank, outputDimensions])
-
-        super.init(
-            weight: linear.weight, bias: linear.bias, scales: linear.scales, biases: linear.biases,
-            groupSize: linear.groupSize, bits: linear.bits)
-
-        // start frozen except for the lora keys
-        freeze()
-    }
-
-    /// Freeze all parameters except the lora parameters
-    public override func freeze(recursive: Bool = true, keys: [String]? = nil, strict: Bool = false)
-        throws
-    {
-        // realize the keys and omit the lora parameters
-        let keys =
-            (keys ?? self.filterMap(filter: Self.filterLocalParameters).flattened().map { $0.0 })
-            .filter {
-                $0 != "lora_a" && $0 != "lora_b"
-            }
-        try super.freeze(recursive: recursive, keys: keys, strict: strict)
-    }
-
-    /// Convert a `QuantizedLinear` layer into a new `Linear` layer
-    /// that implements the `LoRA` adapter.
-    ///
-    /// This is typically called via ``LoRATrain/convert(model:layers:)``.
-    ///
-    /// ### See Also
-    /// - ``LoRATrain/convert(model:layers:)``
-    /// - ``LoRALinear/from(linear:rank:)``
-    public static func from(linear: QuantizedLinear, rank: Int = 8) -> Linear {
-        var (outputDimensions, inputDimensions) = linear.shape
-        inputDimensions = inputDimensions * 32 / linear.bits
-        return QLoRALinear(inputDimensions, outputDimensions, rank: rank, linear: linear)
-    }
-
-    /// Convert back into a fused `QuantizedLinear` layer.
-    ///
-    /// This is typically called via ``LoRATrain/fuse(model:layers:deQuantize:)``.
-    ///
-    /// ### See Also
-    /// - ``LoRATrain/fuse(model:layers:deQuantize:)``
-    public func toLinear(deQuantize: Bool = false) -> Linear {
-        // convert back into full weights
-        let weight = dequantized(
-            weight, scales: scales, biases: biases, groupSize: groupSize, bits: bits)
-
-        let loraB = (scale * loraB.T).asType(.float16)
-        let loraA = loraA.T.asType(.float16)
-
-        // convert back into quantized
-        return QuantizedLinear(
-            weight: weight + matmul(loraB, loraA), bias: bias, groupSize: groupSize, bits: bits)
-    }
-
-    public override func callAsFunction(_ x: MLXArray) -> MLXArray {
-        let y = super.callAsFunction(x.asType(scales.dtype))
-        let z = matmul(matmul(x, self.loraA), self.loraB)
-        return y + scale * z
-    }
-}
-
 /// Equivalent to `lora.py/iterate_batches()`.  Used internally by ``LoRATrain``.
 struct LoRABatchIterator: Sequence, IteratorProtocol {
```

Comment:
@awni I wasn't sure if this training code should stay in LLM or move to LMCommon -- the places where it handles the training input are all around strings, so I think it may need to stay in LLM.